### PR TITLE
Provide a compile option to disable AVX512-specific optimizations.

### PR DESCRIPTION
**Pull request by nebeid**
### Issues:
Addresses P85184645

### Description of changes: 
Describe AWS-LC’s current behavior and how your code changes that behavior. If there are no issues this pr is resolving,
explain why this change is necessary.

### Call-outs:
Point out areas that need special attention or support during the review process. Discuss architecture or design changes.

### Testing:
How is this change tested (unit tests, fuzz tests, etc.)? Are there any testing steps to be verified by the reviewer?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and 
the ISC license.

### CODE REVIEW

File: crypto/fipsmodule/modes/internal.h [hunk 1]
```diff
@@ -280,7 +280,8 @@ void gcm_gmult_ssse3(uint64_t Xi[2], const u128 Htable[16]);
 void gcm_ghash_ssse3(uint64_t Xi[2], const u128 Htable[16], const uint8_t *in,
                      size_t len);
 
-#if defined(OPENSSL_X86_64) && !defined(MY_ASSEMBLER_IS_TOO_OLD_FOR_AVX)
+#if defined(OPENSSL_X86_64) && !defined(MY_ASSEMBLER_IS_TOO_OLD_FOR_AVX) \
+  && !defined(MY_ASSEMBLER_IS_TOO_OLD_FOR_AVX)
```
Comment (dkostic) [inline crypto/fipsmodule/modes/internal.h:284]:
```suggestion
  && !defined(MY_ASSEMBLER_IS_TOO_OLD_FOR_AVX512)
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -224,6 +224,14 @@ if(MY_ASSEMBLER_IS_TOO_OLD_FOR_AVX)
   message(STATUS "MY_ASSEMBLER_IS_TOO_OLD_FOR_AVX selected, removing AVX optimisations")
 endif()
 
+# Some assemblers know about AVX but not AVX512 instructions, e.g. gcc 4.8.2.
+# This flag can be set to handle such cases.
+option(MY_ASSEMBLER_IS_TOO_OLD_FOR_AVX512 "Exclude AVX512 code from the build" OFF)
```
Comment (dkostic) [inline CMakeLists.txt:229]: Having `MY_ASSEMBLER_IS_TOO_OLD_FOR_AVX` set should probably automatically set `MY_ASSEMBLER_IS_TOO_OLD_FOR_AVX512` as well.

File: crypto/fipsmodule/modes/internal.h (hunk 1 above)
Comment (nebeid) [reply crypto/fipsmodule/modes/internal.h:284]: Thank you, I had corrected it on my end but forgot to save it :)

File: tests/ci/run_posix_tests.sh [hunk 3]
```diff
@@ -46,3 +46,8 @@ build_options_to_test=("" "-DBUILD_SHARED_LIBS=1" "-DCMAKE_BUILD_TYPE=Release" "
 for build_option in "${build_options_to_test[@]}"; do
   run_build ${build_option} -DMY_ASSEMBLER_IS_TOO_OLD_FOR_AVX=ON
 done
+
+## Build option: MY_ASSEMBLER_IS_TOO_OLD_FOR_AVX512
+for build_option in "${build_options_to_test[@]}"; do
+  run_build ${build_option} -DMY_ASSEMBLER_IS_TOO_OLD_FOR_AVX512=ON
```
[the remark this answers is not in the document]
Comment (nebeid) [reply tests/ci/run_posix_tests.sh:52]: In the dispatch test, we want to make sure the other functions that have AVX are hit but not the ones with AVX512. Am I on the right track here?

File: CMakeLists.txt (hunk 2 above)
Comment (nebeid) [reply CMakeLists.txt:229]: I thought of that, but we anyway need this section, right? For the cases that disable only AVX512. So NO AVX implies NO AVX512 but not the other way around, so do I just add the AVX512 definition to the section above?
In the AVX512 AESNI perl file, do I just check for AVX512?

File: tests/ci/run_posix_tests.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (dkostic) [reply tests/ci/run_posix_tests.sh:52]: you are right

File: CMakeLists.txt (hunk 2 above)
Comment (dkostic) [reply CMakeLists.txt:230]: > I thought of that, but we anyway need this section, right? 

yes

> do I just add the AVX512 definition to the section above?

yes, you can just set `MY_ASSEMBLER_IS_TOO_OLD_FOR_AVX512` variable to `ON`.

> In the AVX512 AESNI perl file, do I just check for AVX512?

I'd leave both checks, there is no cost to it.

**Comment by nebeid** [comment]
I thought this commit would solve the issue in the build of gcc 4.1.x in our CI, but it seems that the only flag passed is still

**Review by torben-hansen — DISMISSED** [review]

**Review by torben-hansen — APPROVED** [review]

**Review by dkostic — APPROVED** [review]

**Pull request merged**